Devtools darwin

trws opened the pull request:
This does two things:
* add lima to the dev tools stack on linux
* add a devtools stack for Darwin that builds on the PRs of today and yesterday to get a dev environment building consistently on MacOS as well.

Note that this is stacked on top of #46905, that's why there are seemingly non-stack related changes here.  Without them the builds would fail.

trws committed:
stacks: add a stack for devtools on darwin

After getting this whole mess building on darwin, let's keep it that
way, and maybe make it so we have some non-ML darwin binaries in spack
as well.

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/stacks/developer-tools-darwin/spack.yaml:
```diff
@@ -0,0 +1,82 @@
+      require:
+      - target=aarch64
+  concretizer:
+    unify: true
```
If you want things to rebuild when hashes change, I think you want `reuse: false` here too.

trws replied:
Thanks for that, it was wrong on the linux side as well.

alalazo replied:
I think `unify: true` was correct both here and in the linux side. That just means that the environment has a single configuration for each "runtime" package, and can be projected to a standard `/bin`, `/lib` etc. layout without conflicts.

Using `unify: false` we open the door for different specs to use slightly different library dependencies, since the optimization is done spec by spec, instead on the environment as a whole.

alalazo replied:
Fwiw I posted a comment referring to `unify: false`, I saw later that we are discussing `reuse: false` - which is not in the diff. Deleted my previous comment as it was just noise.

scottwittenburg commented:
Might be worth checking with @eugeneswalker whether UO has the capacity to handle these extra builds.

trws commented:
Definitely should.  At the moment the only thing we build on Darwin is an ML stack though, which is less than ideal.

trws committed:
reuse: false for devtools

trws committed:
dtc: fix darwin dylib name and id

On mac the convention is `lib<name>.<num>.dylib`, while the makefile
creates a num suffixed one by default. The id in the file is also a
local name rather than rewritten to the full path, this fixes both
problems.

trws commented:
@scottwittenburg, any idea how I could be getting an error that it couldn't find a spack.lock in the generate phase for the stack? https://gitlab.spack.io/spack/spack/-/jobs/12974881

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/stacks/developer-tools-darwin/spack.yaml:
```diff
@@ -0,0 +1,83 @@
+  concretizer:
+    unify: true
+    reuse: false
+  definitions:
```
@trws I was looking into why there was no lock file to copy to the concrete env dir. I was sort of drawing a blank, but one thing I noticed is that I don't see any messages in the output related to the concretization.  Is it possible your environment is ending up empty? I see here you've got the specs inside `definitions`, but then don't use it in a matrix.  Maybe you can try to remove the `definitions`, and just keep the specs list?

trws replied:
That might well be it, I was trying to refactor and maybe got it wrong, thanks for the advice!

trws committed:
node-js: make whereis more deterministic

trws committed:
relocation(darwin): catch Mach-O load failure

The MachO library can throw an exception rather than return no headers,
this happened in an elf file in the test data of go-bootstrap.  Trying
catching the exception and moving on for now.  May also need to look
into why we're trying to rewrite an elf file.

trws committed:
qemu: add darwin flags to clear out warnings

There's a build failure for qemu in CI, but it's invisible because of
the immense mass of warning output.  Explicitly specify the target macos
version and remove the extraneous unknown-warning-option flag.

trws committed:
dtc: libyaml is also a link dependency

libyaml is required at runtime to run the dtc binary, lack of it caused
the ci for qemu to fail when the library wasn't found.

trws commented:
As of the last push, before the rebase, every package in the stack builds successfully on darwin.  The failure came from two other stacks failing.  Assuming it does the same post-rebase this one may be ready to go.

trws commented:
Ok, everything is passing but codecov. @tgamblin, @eugeneswalker any comments/concerns?

alalazo commented on share/spack/gitlab/cloud_pipelines/stacks/developer-tools-darwin/spack.yaml:
```diff
@@ -0,0 +1,82 @@
+    - libtool
+    - go   # to build fzf, gh, hub
+    - rust+dev   # fd, ripgrep, hyperfine, exa, rust-analyzer
+    # NODARWIN - binutils+ld+gold+plugins   # support linking with built gcc
```
```suggestion
```

alalazo commented on share/spack/gitlab/cloud_pipelines/stacks/developer-tools-darwin/spack.yaml:
```diff
@@ -0,0 +1,82 @@
+      # ensure we can use a jobserver build and do this fast
+    - gmake
+    - ninja   # should be @kitware, can't be because of meson requirement
+    # not for darwin - openssl certs=system     # must be this, system external does not work
```
```suggestion
```

alalazo commented on share/spack/gitlab/cloud_pipelines/stacks/developer-tools-darwin/spack.yaml:
```diff
@@ -0,0 +1,82 @@
+    - ninja   # should be @kitware, can't be because of meson requirement
+    # not for darwin - openssl certs=system     # must be this, system external does not work
+    - libtree
+    # not for darwin - patchelf
```
```suggestion
```

alalazo commented on share/spack/gitlab/cloud_pipelines/stacks/developer-tools-darwin/spack.yaml:
```diff
@@ -0,0 +1,82 @@
+    # not for darwin - patchelf
+    - sed
+    - which
+    # not for darwin - elfutils
+    # not for darwin - fontconfig
+    # not for darwin - font-util
+    # not for darwin - gdb
```
```suggestion
```

alalazo reviewed (dismissed):
LGTM, I just marked some comments for possible cleanup. Feel free to merge if you're not waiting for other reviews.

trws commented:
Updated with those changes @alalazo.
